packages/node-resolve/src/index.js

```diff
@@ -17,16 +17,16 @@ import {
 const builtins = new Set(builtinList);
 const ES6_BROWSER_EMPTY = '\0node-resolve:empty.js';
 const nullFn = () => null;
-const defaults = {
+export const defaults = Object.freeze({
```

Comment:
You're almost there. I mentioned that we needed to **deep freeze** the object exported. Here's why:

```js
> const o = Object.freeze({ batcave: [] })
undefined
> o.batcave.push('joker')
1
> o
{ batcave: [ 'joker' ] }
```

Just using Object.freeze isn't enough here. Users can still modify properties that are themselves objects.

We also don't want to export this directly here. `defaults` should stay internal and we should be able to mutate it as needed internally. The value _exported_ however, should be a deeply frozen copy of this. `Object.freeze` is an in-place method, meaning the return value is the same as the value passed. 

```js
> const a = {}
undefined
> const b = Object.freeze(a)
undefined
> a === b
true
```

So we can't `export` here and we cant `freeze` here either.

Reply:
Can you elaborate why it's important to do a **deep freeze** please?

I can also just do `export const DEFAULTS = { ...defaults };` to become the same I think? 🤔

Reply:
Object.freeze only freezes first-level keys.

This is allowed with shallow freeze, while it shouldn't be with deep freeze:

```
const a = Object.freeze({
  b: 42,
  c: {
    d: 'foo',
  },
});

a.c.d = 'bar';
```

Reply:
> We also don't want to export this directly here. defaults should stay internal and we should be able to mutate it as needed internally. The value exported however, should be a deeply frozen copy of this.

@shellscape, what is the point of mutating defaults internally if exported value will stay the same? That will just lead to more bugs.

Reply:
> That will just lead to more bugs.

It won't. The exported defaults is merely a reference.